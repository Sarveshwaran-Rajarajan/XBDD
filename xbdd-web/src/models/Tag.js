const getStatusPresences = features => {
  const findStatus = status => features.find(feature => feature.calculatedStatus === status);

  return {
    containsPassed: !!findStatus("passed"),
    containsUndefined: !!findStatus("undefined"),
    containsFailed: !!findStatus("failed"),
    containsSkipped: !!findStatus("skipped"),
  };
};

class Tag {
  constructor(data) {
    this.name = data.tag;
    this.features = data.features;
    Object.assign(this, getStatusPresences(data.features));
  }
}

export default Tag;
