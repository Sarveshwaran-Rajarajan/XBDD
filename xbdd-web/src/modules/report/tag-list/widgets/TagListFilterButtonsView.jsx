import React from "react";
import { PropTypes } from "prop-types";
import { withStyles, MuiThemeProvider, createMuiTheme } from "@material-ui/core/styles";
import IconButton from "@material-ui/core/IconButton";
import Tooltip from "@material-ui/core/Tooltip";
import DoneIcon from "@material-ui/icons/Done";
import ErrorOutlineIcon from "@material-ui/icons/ErrorOutline";
import BlockIcon from "@material-ui/icons/Block";
import HelpOutlineIcon from "@material-ui/icons/HelpOutline";
import { tagListFilterButtonStyles } from "../styles/TagListStyles";
import { Card } from "@material-ui/core";

const buttonTheme = color =>
  createMuiTheme({
    palette: {
      primary: { main: color },
      secondary: { main: "#E0E0E0" },
    },
    typography: {
      useNextVariants: true,
    },
  });

const createButton = variant => (
  <MuiThemeProvider theme={variant.theme}>
    <Tooltip title={variant.tooltip} placement="top">
      <IconButton className={variant.className} onClick={variant.handler} variant="outlined" size="small" color={variant.color}>
        {variant.icon}
      </IconButton>
    </Tooltip>
  </MuiThemeProvider>
);

const TagListFilterButtonsView = props => {
  const { classes, filterStates, onFilterButtonClick } = props;
  const variants = {
    passed: {
      theme: buttonTheme("#1d3557"),
      icon: <DoneIcon className={classes.xbddFilterButtonIcon} />,
      tooltip: "Passed",
      color: filterStates.passedSelected ? "primary" : "secondary",
      handler: () => onFilterButtonClick("passedSelected"),
      className: `${classes.xbddFilterButton} ${classes.xbddFilterButtonFirst}`,
    },
    failed: {
      theme: buttonTheme("#E63946"),
      icon: <ErrorOutlineIcon className={classes.xbddFilterButtonIcon} />,
      tooltip: "Failed",
      color: filterStates.failedSelected ? "primary" : "secondary",
      handler: () => onFilterButtonClick("failedSelected"),
      className: `${classes.xbddFilterButton}`,
    },
    undefined: {
      theme: buttonTheme("#A8DADC"),
      icon: <HelpOutlineIcon className={classes.xbddFilterButtonIcon} />,
      tooltip: "Undefined",
      color: filterStates.undefinedSelected ? "primary" : "secondary",
      handler: () => onFilterButtonClick("undefinedSelected"),
      className: `${classes.xbddFilterButton}`,
    },
    skipped: {
      theme: buttonTheme("#457B9D"),
      icon: <BlockIcon className={classes.xbddFilterButtonIcon} />,
      tooltip: "Skipped",
      color: filterStates.skippedSelected ? "primary" : "secondary",
      handler: () => onFilterButtonClick("skippedSelected"),
      className: `${classes.xbddFilterButton} ${classes.xbddFilterButtonLast}`,
    },
  };

  return (
    <Card raised className={classes.xbddTagListFilterButtons}>
      {createButton(variants.passed)}
      {createButton(variants.failed)}
      {createButton(variants.undefined)}
      {createButton(variants.skipped)}
    </Card>
  );
};

TagListFilterButtonsView.propTypes = {
  filterStates: PropTypes.shape({}).isRequired,
  onFilterButtonClick: PropTypes.func.isRequired,
  classes: PropTypes.shape({}).isRequired,
};

export default withStyles(tagListFilterButtonStyles)(TagListFilterButtonsView);
