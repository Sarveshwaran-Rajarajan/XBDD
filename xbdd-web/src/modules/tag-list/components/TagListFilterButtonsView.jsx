import React from 'react';
import { PropTypes } from 'prop-types';
import { withStyles, MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import DoneIcon from '@material-ui/icons/Done';
import ErrorOutlineIcon from '@material-ui/icons/ErrorOutline';
import BlockIcon from '@material-ui/icons/Block';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import { tagListFilterButtonStyles } from '../styles/TagListStyles';

const buttonTheme = color => createMuiTheme({
    palette: {
        primary: { main: color },
        secondary: { main: '#E0E0E0' },
    },
});

const createButton = variant => (
    <MuiThemeProvider theme={variant.theme}>
        <Tooltip title={variant.tooltip} placement="top">
            <IconButton
                className={variant.className}
                onClick={variant.handler}
                variant="outlined"
                size="small"
                color={variant.color}
            >
                {variant.icon}
            </IconButton>
        </Tooltip>
    </MuiThemeProvider>
);

const TagListFilterButtonsView = (props) => {
    const variants = {
        passed: {
            theme: buttonTheme('#1d3557'),
            icon: <DoneIcon />,
            tooltip: 'Passed',
            color: props.filterStates.passedSelected ? 'primary' : 'secondary',
            handler: () => props.onFilterButtonClick('passedSelected'),
            className: `${props.classes.xbddFilterButton} ${props.classes.xbddFilterButtonFirst}`,
        },
        failed: {
            theme: buttonTheme('#E63946'),
            icon: <ErrorOutlineIcon />,
            tooltip: 'Failed',
            color: props.filterStates.failedSelected ? 'primary' : 'secondary',
            handler: () => props.onFilterButtonClick('failedSelected'),
            className: `${props.classes.xbddFilterButton}`,
        },
        undefined: {
            theme: buttonTheme('#A8DADC'),
            icon: <HelpOutlineIcon />,
            tooltip: 'Undefined',
            color: props.filterStates.undefinedSelected ? 'primary' : 'secondary',
            handler: () => props.onFilterButtonClick('undefinedSelected'),
            className: `${props.classes.xbddFilterButton}`,
        },
        skipped: {
            theme: buttonTheme('#457B9D'),
            icon: <BlockIcon />,
            tooltip: 'Skipped',
            color: props.filterStates.skippedSelected ? 'primary' : 'secondary',
            handler: () => props.onFilterButtonClick('skippedSelected'),
            className: `${props.classes.xbddFilterButton} ${props.classes.xbddFilterButtonLast}`,
        },
    };

    return (
        <div className={props.classes.xbddTagListFilterButtons}>
            {createButton(variants.passed)}
            {createButton(variants.failed)}
            {createButton(variants.undefined)}
            {createButton(variants.skipped)}
        </div>
    );
};

TagListFilterButtonsView.propTypes = {
    filterStates: PropTypes.object.isRequired,
    onFilterButtonClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
};

export default withStyles(tagListFilterButtonStyles)(TagListFilterButtonsView);
