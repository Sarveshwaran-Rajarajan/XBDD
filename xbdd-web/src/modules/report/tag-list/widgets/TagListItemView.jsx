import React from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import { ListItem, ListItemIcon, ListItemText, Collapse } from "@material-ui/core";
import { TurnedIn, ExpandLess, ExpandMore } from "@material-ui/icons";
import TagListFeaturesListView from "./TagListFeaturesListView";
import Tag from "../../../../models/Tag";
import { tagListItemStyles } from "../styles/TagListStyles";

const TagListItemView = props => {
  const { tag, isSelected, selectedStatus, handleTagSelect, classes } = props;
  const featureList = tag.features.filter(feature => selectedStatus[feature.calculatedStatus]);
  let className = classes.xbddTagListItemContainer;

  if (isSelected) {
    className += ` ${classes.xbddTagListItemContainerExpanded}`;
  }
  const onClick = () => handleTagSelect(tag.name);

  return (
    <>
      <ListItem button divider onClick={onClick} className={className}>
        <ListItemIcon className={classes.listItemIcon}>
          <TurnedIn />
        </ListItemIcon>
        <ListItemText>{tag.name}</ListItemText>
        {isSelected ? <ExpandLess /> : <ExpandMore />}
      </ListItem>
      <Collapse in={isSelected} timeout="auto" unmountOnExit>
        <TagListFeaturesListView featureList={featureList} />
      </Collapse>
    </>
  );
};

TagListItemView.propTypes = {
  tag: PropTypes.instanceOf(Tag).isRequired,
  isSelected: PropTypes.bool.isRequired,
  selectedStatus: PropTypes.shape({}).isRequired,
  handleTagSelect: PropTypes.func.isRequired,
  classes: PropTypes.shape({}).isRequired,
};

export default withStyles(tagListItemStyles)(TagListItemView);
