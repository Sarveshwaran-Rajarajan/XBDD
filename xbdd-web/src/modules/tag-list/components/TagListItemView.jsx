import React from 'react';
import { PropTypes } from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import Tag from '../../../models/Tag';
import { tagListItemStyles } from '../styles/TagListStyles';

const TagListItemView = (props) => {
    const className = props.isSelected ? `${props.classes.xbddTagListItemContainer} ${props.classes.xbddTagListItemContainerSelected}` : props.classes.xbddTagListItemContainer;
    const onSelectTag = () => props.onSelectTag(props.tag);

    return (
        <ListItem
            button
            onClick={onSelectTag}
            className={className}
        >
            {props.tag.name}
        </ListItem>
    );
};

TagListItemView.propTypes = {
    tag: PropTypes.instanceOf(Tag).isRequired,
    classes: PropTypes.object.isRequired,
    onSelectTag: PropTypes.func.isRequired,
    isSelected: PropTypes.bool.isRequired,
};

export default withStyles(tagListItemStyles)(TagListItemView);
