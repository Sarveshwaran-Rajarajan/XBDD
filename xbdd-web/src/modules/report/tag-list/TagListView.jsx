import React from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import Tag from "../../../models/Tag";
import TagListItemView from "./widgets/TagListItemView";
import TagListFilterButtonsView from "./widgets/TagListFilterButtonsView";
import { tagListStyles } from "./styles/TagListStyles";
import { Card } from "@material-ui/core";

const mapTagToTagListItem = (tag, itemsUIState, filterStates, onSelectTag) => (
  <TagListItemView tag={tag} key={tag.name} itemsUIState={itemsUIState} filterStates={filterStates} onSelectTag={onSelectTag} />
);

const renderList = (className, tags, itemsUIState, filterStates, onSelectTag) => {
  if (!tags.length) {
    return null;
  }
  return (
    <Card raised className={className}>
      <List component="ul">{tags.map(tag => mapTagToTagListItem(tag, itemsUIState[tag.name], filterStates, onSelectTag))}</List>
    </Card>
  );
};

const TagListView = props => (
  <div className={props.classes.xbddTagListContainer}>
    <TagListFilterButtonsView filterStates={props.filterStates} onFilterButtonClick={props.onFilterButtonClick} />
    {renderList(props.classes.xbddTagList, props.tags, props.itemsUIState, props.filterStates, props.onSelectTag)}
  </div>
);

TagListView.propTypes = {
  tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)),
  filterStates: PropTypes.shape({}).isRequired,
  onSelectTag: PropTypes.func,
  onFilterButtonClick: PropTypes.func,
  classes: PropTypes.shape({
    xbddTagListContainer: PropTypes.string,
    xbddTagList: PropTypes.string,
  }).isRequired,
};

TagListView.defaultProps = {
  tags: [],
  selectedTag: null,
  onSelectTag: () => {},
  onFilterButtonClick: () => {},
};

export default withStyles(tagListStyles)(TagListView);
