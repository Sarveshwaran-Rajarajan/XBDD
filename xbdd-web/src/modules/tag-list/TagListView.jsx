import React from "react";
import { PropTypes } from "prop-types";
import { withStyles } from "@material-ui/core/styles";
import List from "@material-ui/core/List";
import Tag from "../../models/Tag";
import TagListItemView from "./widgets/TagListItemView";
import TagListFilterButtonsView from "./widgets/TagListFilterButtonsView";
import { tagListStyles } from "./styles/TagListStyles";

const mapTagToTagListItem = (tag, selectedTag, onSelectTag) => (
    <TagListItemView
        tag={tag}
        key={tag.name}
        onSelectTag={onSelectTag}
        isSelected={tag === selectedTag}
    />
);

const renderList = (tags, selectedTag, onSelectTag) => (
    <List component="ul">
        {tags.map(tag => mapTagToTagListItem(tag, selectedTag, onSelectTag))}
    </List>
);

const TagListView = props => (
    <div className={props.classes.xbddTagListContainer}>
        <TagListFilterButtonsView
            filterStates={props.filterStates}
            onFilterButtonClick={props.onFilterButtonClick}
        />
        <div className={props.classes.xbddTagList}>
            {renderList(props.tags, props.selectedTag, props.onSelectTag)}
        </div>
    </div>
);

TagListView.propTypes = {
    tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)),
    selectedTag: PropTypes.instanceOf(Tag),
    filterStates: PropTypes.shape({
        passedSelected: PropTypes.bool,
        undefinedSelected: PropTypes.bool,
        failedSelected: PropTypes.bool,
        skippedSelected: PropTypes.bool,
    }).isRequired,
    onSelectTag: PropTypes.func,
    onFilterButtonClick: PropTypes.func,
    classes: PropTypes.object.isRequired,
};

TagListView.defaultProps = {
    tags: [],
    selectedTag: null,
    onSelectTag: () => {},
    onFilterButtonClick: () => {},
};

export default withStyles(tagListStyles)(TagListView);
