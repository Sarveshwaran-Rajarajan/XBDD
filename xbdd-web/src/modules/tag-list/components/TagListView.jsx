import React from 'react';
import { PropTypes } from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import Tag from '../../../models/Tag';
import TagListItemView from './TagListItemView';
import TagListFilterButtonsView from './TagListFilterButtonsView';
import { tagListStyles } from '../styles/TagListStyles';

const mapTagToTagListItem = (tag, selectedTag, onSelectTag) => (
    <TagListItemView
        tag={tag}
        key={tag.name}
        onSelectTag={onSelectTag}
        isSelected={tag === selectedTag}
    />);

const renderList = (tags, selectedTag, onSelectTag) => (
    <List component="ul" >
        {tags.map(tag => mapTagToTagListItem(tag, selectedTag, onSelectTag))}
    </List>);

const TagListView = props => (
    <div className={props.classes.xbddTagListContainer}>
        <TagListFilterButtonsView
            filterStates={props.filterStates}
            onFilterButtonClick={props.onFilterButtonClick}
        />
        <div className={props.classes.xbddTagList}>
            {renderList(props.tags, props.selectedTag, props.onSelectTag)}
        </div>
    </div>
);

TagListView.propTypes = {
    tags: PropTypes.arrayOf(PropTypes.instanceOf(Tag)).isRequired,
    selectedTag: PropTypes.instanceOf(Tag),
    filterStates: PropTypes.object.isRequired,
    onSelectTag: PropTypes.func.isRequired,
    onFilterButtonClick: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
};

TagListView.defaultProps = {
    selectedTag: null,
};

export default withStyles(tagListStyles)(TagListView);
