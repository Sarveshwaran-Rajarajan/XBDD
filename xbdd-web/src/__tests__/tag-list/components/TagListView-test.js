import React from "react";
import { shallow, mount } from "enzyme";
import TagListView from "../../../modules/tag-list/components/TagListView";
import TagListItemView from "../../../modules/tag-list/components/TagListItemView";
import TagListFilterButtonsView from "../../../modules/tag-list/components/TagListFilterButtonsView";
import Tag from "../../../models/Tag";

const filterStates = {
    passedSelected: true,
    undefinedSelected: true,
    failedSelected: true,
    skippedSelected: true,
};

const dummyTags = [
    new Tag({
        tag: "testTagOne",
        features: [],
    }),
    new Tag({
        tag: "testTagTwo",
        features: [],
    }),
];

test("TagListView renders", () => {
    const tagList = shallow(<TagListView tags={dummyTags} filterStates={filterStates} />);
    expect(tagList).toMatchSnapshot();
});

test("TagListView renders TagListFilterButtonsView", () => {
    const tagList = mount(<TagListView tags={dummyTags} filterStates={filterStates} />);
    expect(tagList.find(TagListFilterButtonsView)).toHaveLength(1);
});

test("TagListView renders the right number of tags", () => {
    const tagList = mount(<TagListView tags={dummyTags} filterStates={filterStates} />);
    expect(tagList.find(TagListItemView)).toHaveLength(2);
});
