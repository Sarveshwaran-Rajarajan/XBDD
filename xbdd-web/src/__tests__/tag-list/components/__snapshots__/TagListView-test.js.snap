// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagListView renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(TagListView)
    filterStates={
        Object {
            "failedSelected": true,
            "passedSelected": true,
            "skippedSelected": true,
            "undefinedSelected": true,
          }
    }
    tags={
        Array [
            Tag {
              "containsFailed": false,
              "containsPassed": false,
              "containsSkipped": false,
              "containsUndefined": false,
              "features": Array [],
              "name": "testTagOne",
            },
            Tag {
              "containsFailed": false,
              "containsPassed": false,
              "containsSkipped": false,
              "containsUndefined": false,
              "features": Array [],
              "name": "testTagTwo",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "classes": Object {
        "xbddFilterButton": "TagListView-xbddFilterButton-4",
        "xbddTagList": "TagListView-xbddTagList-2",
        "xbddTagListContainer": "TagListView-xbddTagListContainer-1",
        "xbddTagListFilterButtons": "TagListView-xbddTagListFilterButtons-3",
      },
      "filterStates": Object {
        "failedSelected": true,
        "passedSelected": true,
        "skippedSelected": true,
        "undefinedSelected": true,
      },
      "onFilterButtonClick": [Function],
      "onSelectTag": [Function],
      "selectedTag": null,
      "tags": Array [
        Tag {
          "containsFailed": false,
          "containsPassed": false,
          "containsSkipped": false,
          "containsUndefined": false,
          "features": Array [],
          "name": "testTagOne",
        },
        Tag {
          "containsFailed": false,
          "containsPassed": false,
          "containsSkipped": false,
          "containsUndefined": false,
          "features": Array [],
          "name": "testTagTwo",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "classes": Object {
          "xbddFilterButton": "TagListView-xbddFilterButton-4",
          "xbddTagList": "TagListView-xbddTagList-2",
          "xbddTagListContainer": "TagListView-xbddTagListContainer-1",
          "xbddTagListFilterButtons": "TagListView-xbddTagListFilterButtons-3",
        },
        "filterStates": Object {
          "failedSelected": true,
          "passedSelected": true,
          "skippedSelected": true,
          "undefinedSelected": true,
        },
        "onFilterButtonClick": [Function],
        "onSelectTag": [Function],
        "selectedTag": null,
        "tags": Array [
          Tag {
            "containsFailed": false,
            "containsPassed": false,
            "containsSkipped": false,
            "containsUndefined": false,
            "features": Array [],
            "name": "testTagOne",
          },
          Tag {
            "containsFailed": false,
            "containsPassed": false,
            "containsSkipped": false,
            "containsUndefined": false,
            "features": Array [],
            "name": "testTagTwo",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
